### DISCUSSION TITLE
Resolved: Re-exporting types from where they are used?

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1681] [Ideas]
A while ago @thomaseizinger posted about splitting our directory structure from our API [0]. In that discussion, if I remember correctly the idea was presented, to export types from places where they are used, for example since users of the `taproot` module will often use `XOnlyPublicKey` we could re-export the key in the `taproot` module. Fastforward to yesterday and in #1678 I attempted to do such a thing but @Kixunil correctly pointed out that these re-exports make it confusing because the re-exported type can be mistaken for another different type i.e., a dev may see `taproot::XOnlyPublicKey` and wonder if this is a different, more "taprooty" x-only pubkey than the one in secp256k1.

As we attempt to stabalize the API what is the best approach to take?

[0] Mentioning you Thomas in case I misquote you and because I enjoy your opinions on API design.

**Comment by dpc** [comment]
As a user I generally agree that exporting stuff from places (other downstream crates) they are used is kind of  confusing. Especially for items that do (or might) be newtyped for some reason.

If we're talking cross-crates re-exports, then `taproot` crate (if such a crate exists) re-exporting whole `secp256k1` in top level of `taproot` is more manageable.

So then I can:

```rust
use taproot::secp256k1::XOnlyPubKey;
use taproot::something;

...
```

and I can depend on only `taproot`, and it's more obvious that the `taproot::secp256k1` is the `secp256k1`.

**Reply by tcharding** [reply]
Re-stating to see I get it; re-export of `secp256k1` in any module in `rust-bitcoin`  that has types from secp in the public API. Interesting.

**Reply by dpc** [reply]
Just to clarify: I'm not 100% sure if it's a good idea. It makes sense to me and I'm planning to use it. The rationale is: if a `taproot` crate uses `secp256k1::XOnlyPublikKey` somewhere, that makes that `secp256k1` (including its version) part of the `taproot`'s API anyway, so why not expose it and save user the hassle of importing it again?

**Reply by Kixunil** [reply]
Reexporting `secp256k1` from root does make sense if its types are in the public API. I'm OK with reexporting `XOnlyPublicKey` *once* I guess.

One problem with reexports of other crates is that for additional features one has to manually add them to `Cargo.toml` and keep them in sync anyway. Unfortunately, the only solution so far is reexporting all the features which seems messy.

**Comment by Kixunil** [comment]
Not only is it confusing but it also clutters the doc page by having the items at multiple places. As another POV, I never saw such reexports in any crate. At most they reexport things from submodules to supermodules (which is fine).
